### PR TITLE
chore(exporters/elasticsearch-exporter): handle missing ES cluster on open

### LINKED ISSUE #2005: Exporter stream processor fails to start if Elasticsearch exporter fails during open phase
Currently, if the exporter is misconfigured or if the endpoint is unreachable, the broker will fail to open the exporter, which causes the stream processor to fail before starting.

AT:
  - Perform only critical setup in configure/open phases
  - Access ES cluster *only* during export phase

NOTE: there are other issues related to this which might be interesting to revisit: restarting failed services once they die unexpectedly, ignoring exporters which fail during the open phase (such that others can export), etc.

### PR DESCRIPTION
**Pull request by npepinpe**
- moves creating index templates to export phase
- adds fault tolerance test

closes #2005

### CODE REVIEW

**Comment by npepinpe** [comment]
One thing horrible with this PR is that due to the lack of retry strategy in the exporter processor, a missing ES cluster will generate a stupid amount of log as the stream processor keeps retrying the same record over and over again. I'm not sure what is the correct solution here in the scope of this issue though. At least maybe for the integration test we could silence the logger?

**Comment by menski** [comment]
> due to the lack of retry strategy in the exporter processor

do we have a issue for this? If not please create one

File: exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporter.java [hunk 1]
```diff
@@ -137,6 +141,8 @@ private void createIndexTemplates() {
         createValueIndexTemplate(ValueType.WORKFLOW_INSTANCE_SUBSCRIPTION);
       }
     }
+
+    indexTemplatesCreated = true;
```
Comment (Zelldon) [inline exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporter.java]: What happens if the exporter dies on creating `createValueIndexTemplate` for what ever reason and retry it starts again with createRootIndexTemplate etc. Will this break?

File: exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/AbstractElasticsearchExporterIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    assertThat(source).isEqualTo(recordToMap(record));
+  }
+
+  protected ElasticsearchTestClient createElasticsearchClient(
+      ElasticsearchExporterConfiguration configuration) {
+    return new ElasticsearchTestClient(
+        configuration, new ZbLogger("io.zeebe.exporter.elasticsearch"));
+  }
+
+  protected Map<String, Object> recordToMap(final Record<?> record) {
```
Comment (Zelldon) [inline exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/AbstractElasticsearchExporterIntegrationTest.java]: Didn't we have these kind of helper somewhere else?

File: exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterFaultToleranceIT.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    // when
+    exporterBrokerRule.configure("es", ElasticsearchExporter.class, configuration);
+    exporterBrokerRule.start();
+    exporterBrokerRule.publishMessage("message", "123");
+    elastic.start();
+
+    // then
+    esClient = createElasticsearchClient(configuration);
+    final Record<MessageRecordValue> messageRecord =
+        RecordingExporter.messageRecords(MessageIntent.PUBLISHED)
```
Comment (Zelldon) [inline exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterFaultToleranceIT.java]: Do we need to test that also Publish was seen by the exporter?

File: exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterTest.java [hunk 4]
```diff
@@ -45,6 +48,18 @@ public void setUp() {
...
+  @Test
+  public void shouldNotFailOnOpenIfElasticIsUnreachable() {
+    // given
+    final ElasticsearchClient client = spy(new ElasticsearchClient(config, new ZbLogger("test")));
+    final ElasticsearchExporter exporter = createExporter(client);
+    config.index.createTemplate = true;
+
+    // when - then : only fails when trying to export, not before
+    openExporter(exporter);
+    assertThatThrownBy(testHarness::export).isNotNull();
```
Comment (Zelldon) [inline exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterTest.java]: Here we could check for `ElasticsearchExporterException` right?

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Looks good some small remarks, we can talk about them f2f if you like.

File: exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporter.java (hunk 1 above)
Comment (npepinpe) [reply exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporter.java]: No, creating index templates is an upsert operation, so at worse it's doing unnecessary work. This is already the current behavior, in the sense that every time you start the broker we create the templates anyway.

Comment (Zelldon) [reply exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporter.java]: Ok but the data get not lost after recreating?

File: exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterTest.java (hunk 4 above)
Comment (npepinpe) [reply exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterTest.java]: In theory yes, if another error is thrown then you could argue the test isn't testing the correct thing...but on the other hand, if an error is thrown it's also unexpected and it's a bug, so...

But yeah I can add it, it costs nothing to do so.

File: exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporter.java (hunk 1 above)
Comment (npepinpe) [reply exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporter.java]: No, it's a template, it's not the index itself; it just describes how to index records that match the template.

File: exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterFaultToleranceIT.java (hunk 3 above)
Comment (npepinpe) [reply exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterFaultToleranceIT.java]: We test below that this was exported to elastic in the `TestUtil.waitUntil`

File: exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/AbstractElasticsearchExporterIntegrationTest.java (hunk 2 above)
Comment (npepinpe) [reply exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/AbstractElasticsearchExporterIntegrationTest.java]: Maybe? It's marked as an addition, but really I just moved it to the new abstract IT class. You mean somewhere else in the code base?

File: exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporter.java (hunk 1 above)
Comment (Zelldon) [reply exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporter.java]: kk thanks

File: exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/AbstractElasticsearchExporterIntegrationTest.java (hunk 2 above)
Comment (Zelldon) [reply exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/AbstractElasticsearchExporterIntegrationTest.java]: Ah no seems there exist only the to msgpack helpers

File: exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterFaultToleranceIT.java (hunk 3 above)
Comment (Zelldon) [reply exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterFaultToleranceIT.java]: This waits on Published right? But you can't be sure whether it has seen the Publish command or not?

Comment (Zelldon) [reply exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterFaultToleranceIT.java]: But might be tested somewhere else

Comment (npepinpe) [reply exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterFaultToleranceIT.java]: I see what you mean. Yes, it only waits on published, and assumes that publish was seen. And yes, it's tested elsewhere, here it's more that we ensure that we eventually export after elastic appears, but I will add it anyway for completeness.

Comment (Zelldon) [reply exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchExporterFaultToleranceIT.java]: Thanks :+1:

**Comment by npepinpe** [comment]
Updated after feedback

**Review by ChrisKujawa — APPROVED** [review]

**Pull request merged**